README
This is a simple command top copy files from host A to host B (or from A to A).

Installing:
Copy the jar file xfer-xxx.jar to both hosts.
(Java runtime environment is necessary on both hosts).

Running:
There are two modes: receiving and sending. The receiver is a service that waits in a loop for connections from senders and processes the incoming files. The sender is a one-shot program for sending single files or directories (recursively).

Run it on the receiving side like so: "java -jar xfer-20150929.jar" (adjust the name of the .jar file)
And on the sending side like so: "java -jar xfer-20150929.jar -h <receiving host> <file/directory to transfer>"
For command line option help, call "java -jar xfer-20150929.jar -help".
To adjust logging, add the "-l" command line option, e.g. for most detailed: "java -jar xfer-20150929.jar -l SEVERE ... <other options>"

To build, call "build.sh" in a cygwin or Linux environment. Otherwise, just compile the Java classes and run them from where they were built. Or download the .jar file from the github release area.



Changelog
2012-04-03
	incorrect rounding  and poss. div. by zero while dividing for bytesPerSecond: must be
	(transferBytes*1000.0 / (transferTimeMillis)), not 
        (transferBytes / (transferTimeMillis / 1000));

2012-03-28
	receiver returned source md5 instead of locally computed one
	
2012-03-26
	Xfer3: NIO, version xfer3.5

2011-09-13
	Option parser: nicht usage() zeigen, wenn nur Parameter ohne Argument
	uebergeben werden (z.B. "-h")

2011-09-01
	format "Each tick ..." as KiB/MiB/GiB
2011-08-12
	write sent rate in INFO when printing progress bar, else in FINE

2011-08-10
	logging time format ohne millisekunden
	neue Option -mod modify window fuer timestamp-Vergleiche
	Level.FINE statt INFO fuer root directory logging

2011-08-08
	fix formatTransferRate: division gab Ganzzahlwerte, braucht 1024.0 statt 1024

2011-08-05
	Review log levels: default nur Sending.../Receiving... und
	warnungen/fehler; ist übersichtlicher

2011-07-15
	Fix bug: zu sendende Datei im root-Verzeichnis wurde 1. Buchstabe
	abgeschnitten, weil Pfad 1 "/" weniger enthält als bei anderen
	Verzeichnissen
	Fix cygpath replace "/" quoteReplacement"\\"

2011-07-12
	Fix findbugs: DateFormats mit ThreadLocal

2011-07-06
	Fix writeToStream(): schreibe byte[]{0} statt 0 bei compression, sonst u.U.
	ArrayIndexOutOfBoundsException.

2011-06-28 
	Better format transfer rate for sent/received data.
	Fix options error: -b<n> would skip next param

2011-06-27 
	Print progress bar, option "-b[<num>]"

2011-05-16
	Print version in CONFIG log level and with "-v" command line option

2011-03-31
	New option "-i" to restrict incoming ip address ranges

2011-03-29
	Allow empty directories
	New option "-O": if target file exists already, create transferred file
	with suffix ".xfer" because on Windows, BLA.TXT and bla.txt are the same
	file, whereas everywhere else they are not.

2011-03-25
	Stream compression ("-z" option)
	"-Z" <1|5|9> compression grade

2011-03-23
	Fix lastModificationDate on Windows: file must be closed before calling
	method

2011-03-21
	Fix mkdir error if already exists. Remove various flush() calls for
	speedup.
	New class Xfer2 which keeps network connection alive between files; uses
	version "xfer3.3"
	
2011-03-09
	Log modification date of received file, warn if different.


build.sh
#!/bin/sh
set -o errexit
set -o nounset


# modify these:
project=xfer
compile=1
tmpclasses="tmp_classes" # set to "bin" if wished

# no modifications after here:

if [ ! -d src ] ; then
    echo "No src directory"
    exit 1
fi

[ "${OSTYPE}" == "cygwin" ] && isCygwin=1 || isCygwin=0
echo "Running in cygwin: ${isCygwin}"

today=`date +%Y%m%d`
jarfile=${project}-${today}.jar
srcfile=${project}-src-${today}.jar
bindir=bin
jarsdir=jars
exitAfterCompile=0
doClean=0
doJarClean=0
doDistClean=0
doArchive=1
if [ "$#" != 0 ] ; then
    if  [ "$1" == "-h" ] ; then
        echo "build.sh -c          compile only"
        echo "build.sh -clean      clean temporary files"
        echo "build.sh -jarclean  clean jars"
        echo "build.sh -distclean  clean everything but src files"
        echo "build.sh -noarchive  DON'T move built jar files to ${jarsdir}"
        exit 0
    fi
    if  [ "$1" == "-c" ] ; then
        exitAfterCompile=1
    fi
    if  [ "$1" == "-clean" ] ; then
        doClean=1
    fi
    if  [ "$1" == "-jarclean" ] ; then
        doJarClean=1
    fi
    if  [ "$1" == "-distclean" ] ; then
        doDistClean=1
    fi
    if  [ "$1" == "-noarchive" ] ; then
        doArchive=0
    fi
fi
here=`pwd`
cd $here

if [ "${doDistClean}" == "1" ] ; then
    doClean=1
    doJarClean=1
    echo "dist cleaning"
    echo "cleaning BUILD*"
    rm -rf BUILD-* BUILDNO
fi

if [ "${doClean}" == "1" ] ; then
    echo "cleaning dist"
    rm -rf dist
    echo "cleaning dist-src"
    rm -rf dist-src
    echo "cleaning ${tmpclasses}"
    rm -rf ${tmpclasses}
    echo "cleaning sources_list.txt"
    rm -rf  sources_list.txt
fi

if [ "${doJarClean}" == "1" ] ; then
    echo "cleaning jars"
    rm -rf ${project}-src*.jar 
    echo "cleaning source jars  "
    rm -rf ${project}-src*.jar 
fi

if [ "${doDistClean}" == "1" ] || [ "${doClean}" == "1" ] || [ "${doJarClean}" == "1" ] ; then
    exit 0
fi

# also compile, don't take class files from eclipse
echo "compile"
if [ "${compile}" == "1" ] ; then
    find ./src -name *.java > sources_list.txt
    if [ -d ${tmpclasses} ] ; then rm -rf ${tmpclasses} ; fi
    mkdir ${tmpclasses}
    javac -Xlint:deprecation -d ${tmpclasses} -g  @sources_list.txt

    bindir=${tmpclasses}

fi
if [ "${exitAfterCompile}" == "1" ] ; then
    echo "Compile only, exit."
    exit 0
fi

# determine build number
if [ ! -f BUILDNO ] ; then
    echo "1" > BUILDNO
    i=1
else
    i=`cat BUILDNO`
    if [ -f BUILD-${i} ] ; then
        rm BUILD-*
    fi
    (( i ++ ))
fi
echo $i > BUILDNO
touch BUILD-${i}


echo "Build $i"

echo packing bin
cd $here
if [ -d dist ] ; then rm -rf dist ; fi
mkdir dist
rsync -a ${bindir}/ dist
cp -p build.sh BUILDNO BUILD-* dist
cp -p manifest.mf dist
if [ -f Changelog ] ; then cp -p Changelog dist ; fi
cd dist
[ ${isCygwin} ] && \
    jar -cfm `cygpath -w ${here}/${jarfile} manifest.mf .` \
    || \
    jar -cfm ${here}/${jarfile} manifest.mf .

echo packing src
cd $here
if [ -d dist-src ] ; then rm -rf dist-src ; fi
mkdir dist-src
rsync -a src dist-src
for i in  doc/TODO doc/Changelog ; do if [ -f "${i}" ] ; then cp -p "${i}" dist-src ; fi ; done
cp -p manifest.mf build.sh dist-src
[ ${isCygwin} ] && \
    jar -cf `cygpath -w ${here}/${srcfile} dist-src` \
    ||\
    jar -cf ${here}/${srcfile} dist-src

if [ "${doArchive}" = "1" ] ; then
    if [ ! -d ${jarsdir} ] ; then mkdir ${jarsdir} ; fi
    mv ${here}/${jarfile}  ${here}/${jarsdir}
    mv ${here}/${srcfile} ${here}/${jarsdir}
fi


src/DirTree.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Calendar;
import java.util.Collections;
import java.util.Date;
import java.util.List;

/**
* Library to recursively list files.
*  
* @author amhlaobh@users.noreply.github.com
*/
public final class DirTree {

  /**
  * Command line use.
  * @param args - <tt>args[0]</tt> seed directory to start from
  */
  public static void main(String[] args) throws FileNotFoundException {
    if (args.length == 0){
        return;
    }
    File seedDir = new File(args[0]);
    List<File> files = DirTree.getFiles(seedDir);

    //print out all file names, in the the order of File.compareTo()
    for(File file : files ){
        Calendar cal = Calendar.getInstance();
        cal.setTimeInMillis(file.lastModified());
        Date d = cal.getTime();
        System.out.printf("%10d %tH:%tM:%tS %Td.%Tm.%TY %s%n",
                file.length(), 
                d,d,d,d,d,d, 
                file.toString());
    }
  }
  
  /**
  * Recursively walk a directory tree and return a list of all
  * Files found.
  *
  * @param seedDir - the existing directory to start recursing from 
  */
    public static List<File> getFiles(File seedDir) {
        List<File> result =new ArrayList<File>();
        try {
            result = getFileList(seedDir);
        } catch (FileNotFoundException fnfe) {
            System.err.println("Starting directory could not be found: " + seedDir);
        }
        Collections.sort(result);
        return result;
    }

    private static List<File> getFileList(File seedDir)
            throws FileNotFoundException {
        List<File> result = new ArrayList<File>();
        if (!validateDirectory(seedDir)){
            return result;
        }
        File[] filesAndDirs = seedDir.listFiles();
        if (filesAndDirs == null) {
            return result;
        }
        List<File> filesDirsList = Arrays.asList(filesAndDirs);
        for (File file : filesDirsList) {
            if (file.isFile())
                result.add(file); // only add leaf files
              if (file.isDirectory()) {
                // do recursion
                List<File> deeperList = getFileList(file);
                result.add(file);
                result.addAll(deeperList);
            }
        }
        return result;
    }

    /**
     * Directory is valid if it exists, does not represent a file, and can be
     * read.
     */
    private static boolean validateDirectory(File directory)
            throws FileNotFoundException {
        boolean ok = true;
        if (directory == null) {
            System.err.println("Directory should not be null.");
            ok = false;
        } else if (!directory.exists()) {
            System.err.println("Directory does not exist: " + directory);
            ok = false;
        } else if (!directory.isDirectory()) {
            System.err.println("Is not a directory: " + directory);
            ok = false;
        } else if (!directory.canRead()) {
            System.err.println("Directory cannot be read: " + directory);
            ok = false;
        }
        
        return ok;
    }
} 


src/CompressedBlockInputStream.java
/**
 * Input stream that decompresses data.
 *
 * Copyright 2005 - Philip Isenhour - http://javatechniques.com/
 *
 * This software is provided 'as-is', without any express or
 * implied warranty. In no event will the authors be held liable
 * for any damages arising from the use of this software.
 *
 * Permission is granted to anyone to use this software for any
 * purpose, including commercial applications, and to alter it and
 * redistribute it freely, subject to the following restrictions:
 *
 *  1. The origin of this software must not be misrepresented; you
 *     must not claim that you wrote the original software. If you
 *     use this software in a product, an acknowledgment in the
 *     product documentation would be appreciated but is not required.
 *
 *  2. Altered source versions must be plainly marked as such, and
 *     must not be misrepresented as being the original software.
 *
 *  3. This notice may not be removed or altered from any source
 *     distribution.
 *
 * $Id:  1.2 2005/10/26 17:40:19 isenhour Exp $
 * 
 * http://javatechniques.com/blog/compressing-data-sent-over-a-socket/
 *
 * ------
 *  modified by amhlaobh@users.noreply.github.com
 */
import java.io.*;
import java.util.zip.Inflater;
import java.util.zip.DataFormatException;

public class CompressedBlockInputStream extends FilterInputStream {
    /**
     * Buffer of compressed data read from the stream
     */
    private byte[] inBuf = null;

    /**
     * Length of data in the input data
     */
    private int inLength = 0;

    /**
     * Buffer of uncompressed data
     */
    private byte[] outBuf = null;

    /**
     * Offset and length of uncompressed data
     */
    private int outOffs = 0;
    private int outLength = 0;

    /**
     * Inflater for decompressing
     */
    private Inflater inflater = null;

    public CompressedBlockInputStream(InputStream is)
        throws IOException {
        super(is);
        inflater = new Inflater();
    }

    private void readAndDecompress() throws IOException {
        // Read the length of the compressed block
        int ch1 = in.read();
        int ch2 = in.read();
        int ch3 = in.read();
        int ch4 = in.read();
        if ((ch1 | ch2 | ch3 | ch4) < 0)
            throw new EOFException();
        inLength = ((ch1 << 24) + (ch2 << 16) +
            (ch3 << 8) + (ch4 << 0));

        ch1 = in.read();
        ch2 = in.read();
        ch3 = in.read();
        ch4 = in.read();
        if ((ch1 | ch2 | ch3 | ch4) < 0)
            throw new EOFException();
        outLength = ((ch1 << 24) + (ch2 << 16) +
            (ch3 << 8) + (ch4 << 0));

        // Make sure we've got enough space to read the block
        if ((inBuf == null) || (inLength > inBuf.length)) {
            inBuf = new byte[inLength];
        }

        if ((outBuf == null) || (outLength > outBuf.length)) {
            outBuf = new byte[outLength];
        }

        // Read until we're got the entire compressed buffer.
        // read(...) will not necessarily block until all
        // requested data has been read, so we loop until
        // we're done.
        int inOffs = 0;
        while (inOffs < inLength) {
            int inCount =
                in.read(inBuf, inOffs, inLength - inOffs);
            if (inCount == -1) {
                throw new EOFException();
            }
            inOffs += inCount;
        }

        inflater.setInput(inBuf, 0, inLength);
        try {
            inflater.inflate(outBuf);
        }
        catch(DataFormatException dfe) {
            throw new IOException(
                "Data format exception - " +
                dfe.getMessage());
        }

        // Reset the inflator so we can re-use it for the
        // next block
        inflater.reset();

        outOffs = 0;
    }

    public int read() throws IOException {
        if (outOffs >= outLength) {
            try {
                readAndDecompress();
            }
            catch(EOFException eof) {
                return -1;
            }
        }

        return outBuf[outOffs++] & 0xff;
    }

    public int read(byte[] b, int off, int len)
        throws IOException {
        int count = 0;
        while (count < len) {
            if (outOffs >= outLength) {
                try {
                    // If we've read at least one decompressed
                    // byte and further decompression would
                    // require blocking, return the count.
                    if ((count > 0) && (in.available() == 0))
                        return count;
                    else
                        readAndDecompress();
                }
                catch(EOFException eof) {
                    if (count == 0)
                        count = -1;
                    return count;
                }
            }

            int toCopy =
                Math.min(outLength - outOffs, len - count);
            System.arraycopy(
                outBuf, outOffs, b, off + count, toCopy);
            outOffs += toCopy;
            count += toCopy;
        }

        return count;
    }

    public int available() throws IOException {
        // This isn't precise, but should be an adequate
        // lower bound on the actual amount of available data
        return (outLength - outOffs) + in.available();
    }

}


src/CompressedBlockOutputStream.java
/**
 * Output stream that compresses data. A compressed block
 * is generated and transmitted once a given number of bytes
 * have been written, or when the flush method is invoked.
 *
 * Copyright 2005 - Philip Isenhour - http://javatechniques.com/
 *
 * This software is provided 'as-is', without any express or
 * implied warranty. In no event will the authors be held liable
 * for any damages arising from the use of this software.
 *
 * Permission is granted to anyone to use this software for any
 * purpose, including commercial applications, and to alter it and
 * redistribute it freely, subject to the following restrictions:
 *
 *  1. The origin of this software must not be misrepresented; you
 *     must not claim that you wrote the original software. If you
 *     use this software in a product, an acknowledgment in the
 *     product documentation would be appreciated but is not required.
 *
 *  2. Altered source versions must be plainly marked as such, and
 *     must not be misrepresented as being the original software.
 *
 *  3. This notice may not be removed or altered from any source
 *     distribution.
 *
 * $Id:  1.1 2005/10/26 17:19:05 isenhour Exp $
 * 
 * http://javatechniques.com/blog/compressing-data-sent-over-a-socket/
 * 
 * 1. The streams’ compression functionality should be invisible to the application.
 *    In other words, it should be possible to simply wrap them around the socket’s
 *    input and output streams, then write to and read from them without having to 
 *    know when a block of data should be compressed or decompressed.
 * 2. A compressed block of data should be sent whenever a given number of bytes 
 *    have been written. This will prevent OutOfMemoryErrors from happening if a very 
 *    large block of data is written to the stream.
 * 3. The flush method in CompressedBlockOutputStream should cause any buffered 
 *    input to be immediately compressed and transmitted, even if the size threshold
 *    has not been reached yet.
 * 4. Stream meta-data (beyond any header and trailer information used by ZLIB) 
 *    should be kept to a minimum. In practice, the implementation uses 8 bytes of 
 *    header information: One 4-byte integer that indicates the size of the block 
 *    of the compressed data and another 4-byte integer that indicates the size of 
 *    the uncompressed data.
 * 
 * The classes are not thread-safe. Synchronization will be needed if, 
 * for example, multiple threads will be writing to a CompressedBlockOutputStream.
 * 
 * Any buffered data will be sent when the CompressedBlockOutputStream is flushed 
 * or closed. If you need to ensure that buffered data is always sent when the 
 * program terminates, you can use Runtime’s addShutdownHook method to register
 * a thread that will flush the stream even if the program exits uncleanly.
 * 
 * Generally speaking, you will get higher compression levels for larger blocks 
 * of data, since the compression algorithm will have more opportunities to find 
 * patterns and redundancies. The tradeoff, of course, is that larger buffer sizes 
 * increase memory usage and latency.
 *
 * ------
 *  modified by amhlaobh@users.noreply.github.com

 */
import java.io.*;
import java.util.zip.Deflater;

public class CompressedBlockOutputStream extends FilterOutputStream {
    /**
     * Buffer for input data
     */
    private byte[] inBuf = null;

    /**
     * Buffer for compressed data to be written
     */
    private byte[] outBuf = null;

    /**
     * Number of bytes in the buffer
     */
    private int len = 0;

    /**
     * Deflater for compressing data
     */
    private Deflater deflater = null;
    
    //vvv added amhlaobh@users.noreply.github.com
    public CompressedBlockOutputStream(OutputStream os) throws IOException {
        this(os, 1024*16);
    }
    //^^^

    /**
     * Constructs a CompressedBlockOutputStream that writes to
     * the given underlying output stream 'os' and sends a compressed
     * block once 'size' byte have been written. The default
     * compression strategy and level are used.
     */
    public CompressedBlockOutputStream(OutputStream os, int size)
        throws IOException {
        this(os, size,
            Deflater.DEFAULT_COMPRESSION, Deflater.DEFAULT_STRATEGY);
    }

    /**
     * Constructs a CompressedBlockOutputStream that writes to the
     * given underlying output stream 'os' and sends a compressed
     * block once 'size' byte have been written. The compression
     * level and strategy should be specified using the constants
     * defined in java.util.zip.Deflator.
     */
    public CompressedBlockOutputStream(OutputStream os, int size,
        int level, int strategy) throws IOException {
        super(os);
        this.inBuf = new byte[size];
        this.outBuf = new byte[size + 64];
        this.deflater = new Deflater(level);
        this.deflater.setStrategy(strategy);
    }

    protected void compressAndSend() throws IOException {
        //vvv added amhlaobh@users.noreply.github.com
        if (len > inBuf.length){
            System.err.println("ArrayIndexOutOfBoundsException: len="+len+"   inBuf.length="+inBuf.length);
            //return;
        }
        //^^^
        if (len > 0) {
            deflater.setInput(inBuf, 0, len);
            deflater.finish();
            int size = deflater.deflate(outBuf);

            // Write the size of the compressed data, followed
            // by the size of the uncompressed data
            out.write((size >> 24) & 0xFF);
            out.write((size >> 16) & 0xFF);
            out.write((size >>  8) & 0xFF);
            out.write((size >>  0) & 0xFF);

            out.write((len >> 24) & 0xFF);
            out.write((len >> 16) & 0xFF);
            out.write((len >>  8) & 0xFF);
            out.write((len >>  0) & 0xFF);

            out.write(outBuf, 0, size);
            out.flush();

            len = 0;
            deflater.reset();
        }
    }

    public void write(int b) throws IOException {
        //modified amhlaobh@users.noreply.github.com
        //both write methods change len, potentially causing an
        //ArrayIndexOutOfBoundsException, therefore forward to other write
        //method
        //vvv
        write(new byte[]{(byte)b});
        /*
        inBuf[len++] = (byte) b;
        if (len == inBuf.length) {
            compressAndSend();
        }
        */
        //^^^
    }

    public void write(byte[] b, int boff, int blen)
        throws IOException {
        while ((len + blen) > inBuf.length) {
            int toCopy = inBuf.length - len;
            System.arraycopy(b, boff, inBuf, len, toCopy);
            len += toCopy;
            compressAndSend();
            boff += toCopy;
            blen -= toCopy;
        }
        System.arraycopy(b, boff, inBuf, len, blen);
        len += blen;
    }

    public void flush() throws IOException {
        compressAndSend();
        out.flush();
    }

    public void close() throws IOException {
        compressAndSend();
        out.close();
    }
}
